### ISSUE TITLE
Basic repo ops seems WET from the web

### ISSUE THREAD

**Issue by novaugust** [itsgreggreg/umbrella_test#2]
```elixir
Business.Post.get!(id)
Business.Comment.get!(id)
# ^ would require identical code in a lot of places.
# why not just
Business.get!(Post, id)
Business.get!(Comment, id)
```

**Comment by itsgreggreg** [comment]
Not super partial to the interface. I like using the module name in the namespace rather than passing it to the function because there is no direct access the repo, you can only access the records for which there is an implementation. We could also use a macro to DRY it up, I know how much you'd love that. :)

**Comment by novaugust** [comment]
> We could also use a macro to DRY it up

I'd like to look into using behaviours and protocols, but yeah, totally.
```elixir
defmodule Business.Model do
defmacro __using__() do
  quote do
    def get(id), do: Business.Repo.get(__MODULE__, id)
  end
end
end
```

**Comment by novaugust** [comment]
(this is one place where the methods are so simple that yeah, i don't think it's a bad idea ;)

**Comment by novaugust** [comment]
ah, the other concern i have with "pretend there is no repo" is that preloads become really gross (although they always have been?)

one thing we might do is match `Repo.get`  against module names, and if a module has a specific implementation proxy to it
```
defmodule Business.Repo do
  use Ecto.Repo, foo: bar_baz
  def get(Post, id), do: Post.get(id)
  def get(module, id), do: __super(module, id)
```

**Comment by itsgreggreg** [comment]
Yeah i think we're gonna have to figure out what those defaults look like. I'm thinking you can pass `Business.Post.get!` an option that's either a flag like `:all` or `:bare` or something or just pass the things you need to preload as an option.

**Comment by novaugust** [comment]
At some point, I feel like we'd just be making a proxy `Business.Repo` though. A thin veneer reimplementation of it, rather than just exposing it...

**Comment by itsgreggreg** [comment]
Yeah that's why I'm not super positive on the passing of preloads and think we should use a flag system with useful defaults. I totally hear you though, I don't want to basically write a second interface for the Repo. I want to insulate the web from caring about what the Repo does and just expose modules for accessing data. Like what would the interface look like if there was no underlying repo, if we were accessing data from different sources like a sql store or files on disk

**Comment by DCameronMauch** [comment]
I never understood why Ecto didn't automatically load associations at request time, like ActiveRecord does.  Then you would never get error about associations not being loaded.  The downside that if preloads are not done, the dB gets potentially many more queries.

**Comment by itsgreggreg** [comment]
Because of things like our assessment batch. If it preloaded everything
that'd be millions of records just to list the batches. Though they could
have defaulted to preloading allowing you to specify what not to preloaded.
Functional programming?
